apisix/discovery/nacos/factory.lua

```diff
@@ -0,0 +1,293 @@
+    }
+
+    local auth = config.auth or {}
+    -- for backward compat:
+    -- In older method, we passed username and password inside of host
+    -- In new method its passed separately
+    local username, password, host
+    if config.old_conf then
+        -- extract username and password from host
+        host, username, password = get_base_uri(config.hosts)
+    else
+        host = config.hosts[math_random(#config.hosts)]
+        if (auth.username and auth.username ~= "") and (auth.password and auth.password ~= "") then
+            username = auth.username
+            password = auth.password
+         end
+    end
+    if username and username ~= "" and password and password ~= "" then
+        local token = request_login(self, host, username, password)
+        params["accessToken"] = token
+    end
+
+    if auth.token and auth.token ~= "" then
+        params["accessToken"] = auth.token
+    end
+
+    if (auth.access_key and auth.access_key ~= "") and
+       (auth.secret_key and auth.secret_key ~= "") then
+       local ak, data, signature = utils.generate_signature(serv.group_name,
+                                  serv.name, auth.access_key, auth.secret_key)
+       params["ak"] = ak
+       params["data"] = data
+       params["signature"] = signature
+    end
+
+    local instances = request_instance_list(self, params, host)
+    local nodes = {}
+    for _, instance in ipairs(instances) do
+        local node = {
+            host = instance.ip,
+            port = instance.port,
+            weight = instance.weight or self.config.default_weight,
+            metadata = instance.metadata,
+        }
+        -- docs: https://github.com/yidongnan/grpc-spring-boot-starter/pull/496
+        if is_grpc(instance.scheme) and instance.metadata and instance.metadata.gRPC_port then
+            node.port = host.metadata.gRPC_port
+        end
+        core.table.insert(nodes, node)
+    end
+    return nodes
+end
+
+
+local function fetch_full_registry(self)
+    return function (premature)
+        if premature then
+            return
+        end
+
+        local config = self.config
+        local services_in_use = utils.get_nacos_services(config.id)
+        local service_names = {}
+        for _, serv in ipairs(services_in_use) do
+            if self.stop_flag then
+                core.log.error("nacos client is exited, id: ", config.id)
+                return
+            end
+            local key = utils.generate_key(serv.id, serv.namespace_id,
+                                          serv.group_name, serv.name)
+            service_names[key] = true
+            local nodes = self:fetch_instances(serv)
+            if #nodes > 0 then
+                local content = core.json.encode(nodes)
+                local key = key
+                nacos_dict:set(key, content)
+             end
+        end
+        -- delete unused service names
+        local keys = nacos_dict:get_keys(0)
```

Comment:
> CAUTION Avoid calling this method on dictionaries with a very large number of keys as it may lock the dictionary for significant amount of time and block Nginx worker processes trying to access the dictionary.

According to the OpenResty documentation, I realized that `get_keys` should be avoided. We can record `services_in_use` in a module variable so that by comparing the values of `services_in_use` before and after executing the `fetch_full_registry` function, we can determine the changes without using `get_keys`.